### ISSUE TITLE
Completer closes when typing a letter after dot in Python

### ISSUE THREAD

**Issue by phil-zxx** [jupyterlab/jupyterlab#11377]
When using code completion in **JupyterLab**, it currently works as follows:

- **Image 1**: You type e.g. `os.` and then press `TAB` to get all available suggestions. So far so good.
- **Image 2**: You press another letter e.g. `g` to further filter the suggestions. Unfortunately, the dropdown menu just disappears.
- **Image 3**: You need to press `TAB` a **second time** to get the filtered suggestions back.

![image](https://user-images.githubusercontent.com/32247590/139522440-3aae492e-ad40-4b4d-9978-9866814c9aa2.png)

I believe this behaviour was different a few versions back. Could we revert to the (I'd argue) "correct" behaviour, where the dropdown menu does not disappear, i.e. one has to press `TAB` only once?

---

System Details:
- Windows 10
- Python 3.8.7
- JupyterLab 3.2.1

**Comment by krassowski** [comment]
It is an edge case. In other instances the completer does not close when typing (I will update the title to correct that). This one is tricky because we have a change of tokens; `os.path` gets tokenized (by CodeMirror Python mode) as `os`, `.` and `path` so the completer does not recognize the letter typed after `os.` as belonging to the same continuum (which should get fixed).

**Comment by phil-zxx** [comment]
Thanks for the quick response.

However, I'd argue this is not an edge case, because I know multiple people who have this problem. And I have this problem at home as well as at work, which are 2 independent machines.

**Comment by krassowski** [comment]
I did not mean to say that this is unimportant edge case. This is the kind of an edge case that I woke up at 9 am on Saturday to investigate. Bear we me while I look into this please.

**Comment by krassowski** [comment]
This looks like a regression introduced in 3.1, possibly with RTC. It seems that the closure comes from `handleCursorChange()`; it looks like it gets triggered differently than it used to. The following logic:

https://github.com/jupyterlab/jupyterlab/blob/d48e0c04efb786561137fb20773fc15788507f0a/packages/completer/src/model.ts#L312-L322

gets a `cursorDelta` of zero and `inputDelta` of zero which leads it to (incorrectly) conclude that the cursor has moved beyond the completed token. It looks like a CodeMirror wrapper now emits the signal in a wrong order (or does not emit text changed signal) and we are getting incorrect value for the `currentLine`.

The `currentLine` gets obtained from `this.current` which in turns is set in `handleTextChange`:

https://github.com/jupyterlab/jupyterlab/blob/d48e0c04efb786561137fb20773fc15788507f0a/packages/completer/src/model.ts#L328-L344

the problem is that in `os.<tab>a` case `handleTextChange` does not get invoked before `handleCursorChange`. So both the old line and current line are `os.`.

Neither `onSelectionsChanged`, `handleCursorChange`, nor `handleTextChange` was modified in last 3-4 years. This leads me to believe that this might be related to a very subtle change in the order of events emitted by native CodeMirror and y.js CodeMirror wrapper. Any ideas @dmonad?

**Comment by krassowski** [comment]
This is what I gather, please correct me if I'm wrong. The `onSelectionsChanged` (`handleCursorChange`) and `onTextChanged` (`handleTextChange`) events are triggered by changes to observable elements of the model:

https://github.com/jupyterlab/jupyterlab/blob/6631799510a90938a0ed8bc6fd02b0fc55a067a0/packages/completer/src/handler.ts#L111-L112

which are `IObservableString` and `IObservableMap` respectively (`ModelDB`); the selections get set by public `setCursorPosition` and more importantly private `_onCursorActivity`:

https://github.com/jupyterlab/jupyterlab/blob/f07934a8d564ea5a89ec3b8d681a29115a5d4547/packages/codemirror/src/editor.ts#L875-L882

which is connected to CodeMirror in `cursorActivity` event:

https://github.com/jupyterlab/jupyterlab/blob/f07934a8d564ea5a89ec3b8d681a29115a5d4547/packages/codemirror/src/editor.ts#L168

and also invoked on editor focus:

https://github.com/jupyterlab/jupyterlab/blob/f07934a8d564ea5a89ec3b8d681a29115a5d4547/packages/codemirror/src/editor.ts#L1040-L1050

while the `value` used to be modified by the editor with `_beforeDocChanged` prior to the change happening (on `beforeChange`) thus making it possible to get the new value in `handleCursorChange`:

https://github.com/jupyterlab/jupyterlab/blob/f07934a8d564ea5a89ec3b8d681a29115a5d4547/packages/codemirror/src/editor.ts#L989-L1009

but it is no longer invoked with the code essentially commented out by `y.js`:

https://github.com/jupyterlab/jupyterlab/blob/f07934a8d564ea5a89ec3b8d681a29115a5d4547/packages/codemirror/src/editor.ts#L169-L173

Instead the value updates are now handled in:

https://github.com/jupyterlab/jupyterlab/blob/0e42117a1dcaa9e2980cedca45513ea0db061ec9/packages/codeeditor/src/editor.ts#L282-L302

@dmonad, @hbcarlos can we somehow restore the old sequence of events to reliably have the text update prior to the cursor event? I imagine we would need to delay emitting `_onCursorActivity` after getting it from CodeMirror's `cursorActivity` event up until the model has been updated. Does it make sense? Is there a promise that we could await to do so?

As a proof-of-concept, modifying the `cursorActivity` handler to put `_onCursorActivity` invokation it at the end of the execution queue:

```js
    CodeMirror.on(editor, 'cursorActivity', () => {
      window.setTimeout(() => {
        this._onCursorActivity();
      }, 0);
      }
    );
```

solves the issue for me. But I wonder if we could have a cleaner solution. Sorry about pinging you on Saturday!

PS. I think that this might be also the cause of some other issues with completer I experience since 3.1, it would be important to fix it.

**Comment by krassowski** [comment]
Hi @phil-zxx would you mind checking if the fix in https://github.com/jupyterlab/jupyterlab/pull/11379 works well for you on the [binder](https://mybinder.org/v2/gh/krassowski/jupyterlab/fix-completer?urlpath=lab-dev)?

**Comment by phil-zxx** [comment]
Hi @krassowski. When I tried your PR on binder, there were no issues. As a cross check, I tried the binder of another (random) PR, and the bug was still present there. So I can confirm that your PR fixed the issue!

Thanks so much for the quick turn around. This had been bugging me for quite some time!

**Comment by dmonad** [comment]
Thanks for the ticket @phil-zxx and thanks for the quick PR @krassowski!

@krassowski Let's try to restore the order of the events instead of going the setTimeout approach. This would definitely break if two characters are inserted in the same execution run (e.g. `cell.insert(0, 'a'); cell.insert(1, 'b')`) because the position of the cursor event will then be broken.

Carlos and me are going to look into this.

**Comment by phil-zxx** [comment]
Thanks @krassowski and @dmonad for making this available in 3.2.x! 4.0 would have been a long wait. :)
